linter/ginkgo_linter.go

```diff
@@ -498,6 +507,116 @@ func checkMatchErrorAssertion(pass *analysis.Pass, matcher *ast.CallExpr) (bool,
+//
+// but these are not
+//
+//	// function signature is `os.ReadFile(string) ([]byte, error)`
+//	Expect(err).To(Succeed())
+//	Expect(return os.ReadFile("someFile").To(Succeed())
+func checkSucceedSync(
+	pass *analysis.Pass, assertionExp *ast.CallExpr, actualArgs []ast.Expr, handler gomegahandler.Handler, reportBuilder *reports.Builder,
+	suppressStyleIssue bool,
+) bool {
+	_, isMatcherFuncCall := assertionExp.Args[0].(*ast.CallExpr)
+	if !isMatcherFuncCall {
+		return false
+	}
+
+	assertionExpCopy := astcopy.CallExpr(assertionExp)
+	matcherCopy := assertionExpCopy.Args[0].(*ast.CallExpr)
+
+	return doCheckSucceed(
+		pass, assertionExpCopy, matcherCopy, actualArgs, handler, reportBuilder, false,
+		suppressStyleIssue,
+	)
+}
+
+// Ensure that the Succeed matcher is only used against functions that return a single error
+// in the asynchronous case.
+//
+// e.g. these are valid assertions
+//
+//	// function signature is `os.Remove(string) error`
+//	Eventually(func() error { return os.Remove("emptyDirectory") }).Should(Succeed())
+//	Consistently(func() error { return os.Remove("nonemptyDirectory") }).ShouldNot(Succeed())
+//
+// but this is not
+//
+//	// function signature is `os.ReadFile(string) ([]byte, error)`
+//	Eventually(func() ([]byte, error) { return os.ReadFile("someFile") }).Should(Succeed())
+func checkSucceedAsync(
+	pass *analysis.Pass, assertionExp *ast.CallExpr, actualArgs []ast.Expr, handler gomegahandler.Handler, reportBuilder *reports.Builder,
+	suppressWarnings bool,
+) bool {
+	_, isMatcherFuncCall := assertionExp.Args[0].(*ast.CallExpr)
+	if !isMatcherFuncCall {
+		return false
+	}
+
+	assertionExpCopy := astcopy.CallExpr(assertionExp)
+	matcherCopy := assertionExpCopy.Args[0].(*ast.CallExpr)
+
+	return doCheckSucceed(pass, assertionExpCopy, matcherCopy, actualArgs, handler, reportBuilder, true, suppressWarnings)
+}
+
+// Function that looks for Succeed matcher by recursively unwrapping the `matcher` argument and,
+// if a match is found, enforces rules for the Succeed matcher, i.e. that the actual arguments must
+// be a single error or function returning a single error, depending on whether the assertion is
+// async.
+func doCheckSucceed(
+	pass *analysis.Pass,
+	assertionExpr *ast.CallExpr, matcher *ast.CallExpr, actualArgs []ast.Expr, handler gomegahandler.Handler, reportBuilder *reports.Builder,
+	isAsync, suppressWarnings bool,
+) bool {
+	matcherFuncName, isMatcherFuncCall := handler.GetActualFuncName(matcher)
+	if !isMatcherFuncCall {
+		return false
+	}
+
+	switch matcherFuncName {
+	case succeed:
+	case not:
+		nested, ok := matcher.Args[0].(*ast.CallExpr)
+		if !ok {
+			return false
+		}
+
+		return doCheckSucceed(pass, assertionExpr, nested, actualArgs, handler, reportBuilder, isAsync, suppressWarnings)
+	default:
+		return false
+	}
+
+	if isAsync {
+		if len(actualArgs) != 1 ||
+			!(isExprErrFunc(pass, actualArgs[0]) || isExprGomegaFunc(pass, actualArgs[0])) {
```

Comment:
the function under test is not always the first argument. e.g. this code is valid:

```go
Eventually(ctx, func(...){...}).Should(Succeed())
```

Eventually is with very flexible api...

Reply:
This is actually checked here: https://github.com/nunnatsa/ginkgolinter/blob/main/linter/ginkgo_linter.go#L743-L759